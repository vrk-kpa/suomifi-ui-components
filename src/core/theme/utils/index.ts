export {
  withSuomifiDefaultProps,
  themeOrTokens,
  TokensOrThemeProps,
} from './defaults';
export { focus } from './focus';
export { boxshadowOutline } from './outline';
export { margin, padding, spacingModifiers } from './spacing';
export { typographyUtils, fontSize } from './typography';
