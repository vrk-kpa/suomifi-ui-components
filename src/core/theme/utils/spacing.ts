import {
  SuomifiTheme,
  SpacingProp,
  defaultSuomifiTheme,
} from '../SuomifiTheme/SuomifiTheme';

export type SpacingWithoutInsetProp =
  | 'xxs'
  | 'xs'
  | 's'
  | 'm'
  | 'l'
  | 'xl'
  | 'xxl'
  | 'xxxl'
  | 'xxxxl'
  | '0';

export type GlobalMarginProps = { globalMargins: GlobalMargins };

const spaceVal = (theme: SuomifiTheme) => (val?: SpacingProp) => {
  if (val === '0') return '0';
  return !!val ? theme.spacing[val] : '';
};

export interface PaddingProps {
  /** Padding from theme */
  padding?: SpacingWithoutInsetProp;
  /** Padding-top from theme */
  pt?: SpacingWithoutInsetProp;
  /** Padding-right from theme */
  pr?: SpacingWithoutInsetProp;
  /** Padding-bottom from theme */
  pb?: SpacingWithoutInsetProp;
  /** Padding-left from theme */
  pl?: SpacingWithoutInsetProp;
  /** Padding on the x-axis (left & right) from theme */
  px?: SpacingWithoutInsetProp;
  /** Padding on the y-axis (top & bottom) from theme */
  py?: SpacingWithoutInsetProp;
}
export interface MarginProps {
  /** Margin from theme */
  margin?: SpacingWithoutInsetProp;
  /** Margin-top from theme */
  mt?: SpacingWithoutInsetProp;
  /** Margin-right from theme */
  mr?: SpacingWithoutInsetProp;
  /** Margin-bottom from theme */
  mb?: SpacingWithoutInsetProp;
  /** Margin-left from theme */
  ml?: SpacingWithoutInsetProp;
  /** Margin on the x-axis (left & right) from theme */
  mx?: SpacingWithoutInsetProp;
  /** Margin on the y-axis (top & bottom) from theme */
  my?: SpacingWithoutInsetProp;
}

export type GlobalMargins = {
  all?: MarginProps;
  alert?: MarginProps;
  actionMenu?: MarginProps;
  block?: MarginProps;
  button?: MarginProps;
  breadcrumb?: MarginProps;
  checkbox?: MarginProps;
  checkboxGroup?: MarginProps;
  chip?: MarginProps;
  dateInput?: MarginProps;
  dropdown?: MarginProps;
  expander?: MarginProps;
  expanderGroup?: MarginProps;
  externalLink?: MarginProps;
  fileInput?: MarginProps;
  heading?: MarginProps;
  hintText?: MarginProps;
  inlineAlert?: MarginProps;
  label?: MarginProps;
  languageMenu?: MarginProps;
  link?: MarginProps;
  linkList?: MarginProps;
  loadingSpinner?: MarginProps;
  multiSelect?: MarginProps;
  notification?: MarginProps;
  pagination?: MarginProps;
  paragraph?: MarginProps;
  radioButton?: MarginProps;
  radioButtonGroup?: MarginProps;
  routerLink?: MarginProps;
  searchInput?: MarginProps;
  serviceNavigation?: MarginProps;
  sideNavigation?: MarginProps;
  singleSelect?: MarginProps;
  staticChip?: MarginProps;
  statusText?: MarginProps;
  text?: MarginProps;
  textarea?: MarginProps;
  textInput?: MarginProps;
  timeInput?: MarginProps;
  toast?: MarginProps;
  toggleInput?: MarginProps;
  toggleButton?: MarginProps;
  tooltip?: MarginProps;
  wizardNavigation?: MarginProps;
};

export interface SpacingProps extends PaddingProps, MarginProps {}

export const spacingStyles = (props: SpacingProps | undefined) => {
  if (!props) return;
  const array = Object.entries(props).map(([key, value]) =>
    getSpacingStyle(defaultSuomifiTheme, key as keyof SpacingProps, value),
  );
  return Object.assign({}, ...array);
};

export const buildSpacingCSS = (
  spacing?: SpacingProps,
  important?: boolean,
): string => {
  if (!spacing) return '';

  const cssStyles = Object.entries(spacing)
    .map(([key, value]) =>
      getCSSSpacing(
        defaultSuomifiTheme,
        key as keyof SpacingProps,
        value,
        important,
      ),
    )
    .join('');

  return cssStyles;
};

const inlineStyle = {
  margin: 'margin',
  mt: 'marginTop',
  mr: 'marginRight',
  mb: 'marginBottom',
  ml: 'marginLeft',
  padding: 'padding',
  pt: 'paddingTop',
  pr: 'paddingRight',
  pb: 'paddingBottom',
  pl: 'paddingLeft',
};

const cssSelector = {
  margin: 'margin',
  mt: 'margin-top',
  mr: 'margin-right',
  mb: 'margin-bottom',
  ml: 'margin-left',
  padding: 'padding',
  pt: 'padding-top',
  pr: 'padding-right',
  pb: 'padding-bottom',
  pl: 'padding-left',
};

const getSpacingStyle = (
  theme: SuomifiTheme,
  key: keyof SpacingProps,
  value: SpacingProp,
) => {
  const amount = spaceVal(theme)(value);
  switch (key) {
    case 'mx':
      return { marginRight: `${amount}`, marginLeft: `${amount}` };
    case 'my':
      return { marginTop: `${amount}`, marginBottom: `${amount}` };
    case 'px':
      return { paddingRight: `${amount}`, paddingLeft: `${amount}` };
    case 'py':
      return { paddingTop: `${amount}`, paddingBottom: `${amount}` };
    case 'margin':
    case 'mt':
    case 'mr':
    case 'mb':
    case 'ml':
    case 'padding':
    case 'pt':
    case 'pr':
    case 'pb':
    case 'pl':
      return { [inlineStyle[key]]: `${amount}` };
    default:
      return '';
  }
};

const getCSSSpacing = (
  theme: SuomifiTheme,
  key: keyof SpacingProps,
  value: SpacingProp,
  important: boolean = false,
) => {
  const amount = spaceVal(theme)(value);
  const importantValue = important ? '!important' : '';
  switch (key) {
    case 'mx':
      return `margin-right: ${amount} ${importantValue}; margin-left: ${amount} ${importantValue};`;
    case 'my':
      return `margin-top: ${amount} ${importantValue}; margin-bottom: ${amount} ${importantValue};`;
    case 'px':
      return `padding-right: ${amount} ${importantValue}; padding-left: ${amount} ${importantValue};`;
    case 'py':
      return `padding-top: ${amount} ${importantValue}; padding-bottom: ${amount} ${importantValue};`;
    default:
      return `${cssSelector[key]}: ${amount} ${importantValue};`;
  }
};

export const separateMarginProps = <T extends MarginProps>(
  props: T,
): [MarginProps, Omit<T, keyof MarginProps>] => {
  const { margin, my, mx, mt, mr, mb, ml, ...otherProps } = props;
  return [
    {
      ...(margin !== undefined && { margin }),
      ...(my !== undefined && { my }),
      ...(mx !== undefined && { mx }),
      ...(mt !== undefined && { mt }),
      ...(mr !== undefined && { mr }),
      ...(mb !== undefined && { mb }),
      ...(ml !== undefined && { ml }),
    },
    otherProps,
  ];
};

export const separateMarginAndPaddingProps = <T extends SpacingProps>(
  props: T,
): [SpacingProps, Omit<T, keyof SpacingProps>] => {
  const {
    margin,
    padding,
    my,
    mx,
    py,
    px,
    mt,
    mr,
    mb,
    ml,
    pt,
    pr,
    pb,
    pl,
    ...otherProps
  } = props;
  return [
    {
      ...(margin !== undefined && { margin }),
      ...(padding !== undefined && { padding }),
      ...(my !== undefined && { my }),
      ...(mx !== undefined && { mx }),
      ...(py !== undefined && { py }),
      ...(px !== undefined && { px }),
      ...(mt !== undefined && { mt }),
      ...(mr !== undefined && { mr }),
      ...(mb !== undefined && { mb }),
      ...(ml !== undefined && { ml }),
      ...(pt !== undefined && { pt }),
      ...(pr !== undefined && { pr }),
      ...(pb !== undefined && { pb }),
      ...(pl !== undefined && { pl }),
    },
    otherProps,
  ];
};
