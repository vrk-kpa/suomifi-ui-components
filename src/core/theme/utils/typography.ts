import {
  cssObjectsToCss,
  FlattenSimpleInterpolation,
  cssValueToString,
} from '../../../utils/css';
import { tokens, TypograhpyDesingTokens } from 'suomifi-design-tokens';
import { SuomifiTheme } from '../';

export type TypographyTokensAsCss = {
  [key in keyof typeof tokens.typography]: string
};

type TypographyTokensAsCssProp = {
  [key in keyof TypograhpyDesingTokens]: FlattenSimpleInterpolation
};
export interface TypographyUtil extends TypographyTokensAsCssProp {}
export class TypographyUtil {
  static instance: TypographyUtil;
  constructor(tokens: TypograhpyDesingTokens) {
    // If instance not created, does not take on account if tokens is different!
    if (!TypographyUtil.instance) {
      // Assing typographyTokens as CSS FlattenSimpleInterpolation to this object
      Object.assign(this, cssObjectsToCss(tokens));
      TypographyUtil.instance = this;
    }
    return TypographyUtil.instance;
  }
}

/**
 * Add font-declarations to typography tokens
 * @param typography Typography-tokens
 * @return typographyWithUtils Typography-tokens and typography-tokens as CSS strings
 */
export const typographyUtils = (typography: TypograhpyDesingTokens) => {
  const instance = new TypographyUtil(typography);
  Object.freeze(instance);
  return instance;
};

type ValueTypographyTokenProp = keyof typeof tokens.typography;
export const fontSize = (props: {
  theme: SuomifiTheme;
  [key: string]: any | any[];
}) => (typographyToken: ValueTypographyTokenProp) => {
  return cssValueToString(
    props.theme.values.typography[typographyToken].fontSize,
  );
};
