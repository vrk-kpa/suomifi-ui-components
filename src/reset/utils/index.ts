import { css } from 'styled-components';

const normalize = {
  html:
    'line-height: 1.15; -ms-text-size-adjust: 100%; -webkit-text-size-adjust: 100%;',
  h1: 'font-size: 2em; margin: 0.67em 0;',
  'dl dl': 'margin: 0;',
  'dl ol': 'margin: 0;',
  'dl ul': 'margin: 0;',
  'ol dl': 'margin: 0;',
  'ul dl': 'margin: 0;',
  'ol ol': 'margin: 0;',
  'ol ul': 'margin: 0;',
  'ul ol': 'margin: 0;',
  'ul ul': 'margin: 0;',
  hr: 'box-sizing: content-box; height: 0; overflow: visible;',
  main: 'display: block;',
  pre: 'font-family: monospace, monospace; font-size: 1em;',
  a: 'background-color: transparent;',
  'abbr[title]':
    'text-decoration: underline; text-decoration: underline dotted;',
  b: 'font-weight: bolder;',
  strong: 'font-weight: bolder;',
  code: 'font-family: monospace, monospace; font-size: 1em;',
  kbd: 'font-family: monospace, monospace; font-size: 1em;',
  samp: 'font-family: monospace, monospace; font-size: 1em;',
  small: 'font-size: 80%;',
  audio: 'display: inline-block;',
  video: 'display: inline-block;',
  'audio:not([controls])': 'display: none; height: 0;',
  img: 'border-style: none;',
  'svg:not(:root)': 'overflow: hidden;',
  button:
    'margin: 0;overflow: visible; text-transform: none;-webkit-appearance: button;',
  input: 'margin: 0;overflow: visible;',
  select: 'margin: 0;text-transform: none;',
  '[type="button"]': '-webkit-appearance: button;',
  '[type="reset"]': '-webkit-appearance: button;',
  '[type="submit"]': '-webkit-appearance: button;',
  fieldset: 'padding: 0.35em 0.75em 0.625em;',
  legend:
    'box-sizing: border-box; color: inherit; display: table; max-width: 100%; white-space: normal;',
  progress: 'display: inline-block; vertical-align: baseline;',
  textarea: 'margin: 0; overflow: auto;',
  '[type="checkbox"]': 'box-sizing: border-box; padding: 0;',
  '[type="radio"]': 'box-sizing: border-box; padding: 0;',
  '[type="search"]': '-webkit-appearance: textfield; outline-offset: -2px;',
  '::-webkit-inner-spin-button': 'height: auto;',
  '::-webkit-outer-spin-button': 'height: auto;',
  '::-webkit-input-placeholder': 'color: inherit; opacity: 0.54;',
  '::-webkit-search-decoration': '-webkit-appearance: none;',
  '::-webkit-file-upload-button': '-webkit-appearance: button; font: inherit;',
  '::-moz-focus-inner': 'border-style: none; padding: 0;',
  ':-moz-focusring': 'outline: 1px dotted ButtonText;',
  ':-moz-ui-invalid': 'box-shadow: none;',
  details: 'display: block;',
  dialog:
    // eslint-disable-next-line max-len
    'background-color: white; border: solid; color: black; display: block; height: -moz-fit-content; height: -webkit-fit-content; height: fit-content; left: 0; margin: auto; padding: 1em; position: absolute; right: 0; width: -moz-fit-content; width: -webkit-fit-content; width: fit-content;',
  'dialog:not([open])': 'display: none;',
  summary: 'display: list-item;',
  canvas: 'display: inline-block;',
  template: 'display: none;',
  '[hidden]': 'display: none;',
} as Record<string, string>;

const common = css`
  margin: 0;
  padding: 0;
  border: 0;
  box-sizing: border-box;
  font: 100% inherit;
  line-height: 1;
  text-align: left;
  text-decoration: none;
  vertical-align: baseline;
  color: inherit;
  background: none;
  cursor: inherit;

  ::before,
  ::after {
    box-sizing: border-box;
  }
`;

const resets = {
  normalize,
  common,
};

const selectorDeclaration = (selector: string, styles: string) =>
  `&${selector} {
    ${styles}
  }`;

const selectorsAndCss = (selector: string) => {
  const styles = !!normalize && normalize[selector];
  return !!styles ? selectorDeclaration(selector, styles) : '';
};

/**
 * Return selectors with resets in Emotion-css
 * @param selectors string or array of strings
 */
const resetWithSelectors = (selectors: string | string[]): string => {
  if (Array.isArray(selectors) && selectors.length > 0) {
    return `
      ${selectors.map((s) => selectorsAndCss(s)).join('')}
    `;
  }
  if (typeof selectors === 'string') {
    return selectorsAndCss(selectors);
  }
  return '';
};

export { resets, resetWithSelectors };

// TODO: Add tests
// TODO: create versioned interface for normalize.css
