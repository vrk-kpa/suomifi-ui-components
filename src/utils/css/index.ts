export { alphaHex, hslaToHex } from './colors';
export { allStates, absolute } from './pseudo';
export {
  clearfix,
  cssObjectToCss,
  cssObjectsToCss,
  FlattenSimpleInterpolation,
  cssValueToString,
} from './utils';
