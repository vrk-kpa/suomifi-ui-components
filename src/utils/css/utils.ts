import { css, FlattenSimpleInterpolation } from 'styled-components';
import { ValueUnit } from 'suomifi-design-tokens';
export { FlattenSimpleInterpolation };

export const clearfix = css`
  &:after {
    display: block;
    content: '';
    clear: both;
  }
`;

/**
 * Return FlattenSimpleInterpolation compatible string or number
 * @param cssValue compatible value or {value, unit} pair-object
 */
const cssValueToString = (cssValue: number | string | ValueAndUnit) => {
  if (!!cssValue && typeof cssValue === 'object' && 'value' in cssValue) {
    const { value, unit } = cssValue;
    return !!unit ? `${value}${unit}` : value;
  }
  return cssValue;
};

const camelToSnake = (string: string) =>
  string.replace(/[\w]([A-Z])/g, m => `${m[0]}-${m[1]}`).toLowerCase();
/**
 * Convert CSSObject to CSS FlattenSimpleInterpolation
 * @param value CSSObject
 */
export const cssObjectToCss = <T>(value: T) => css`
  ${Object.entries(value)
    .map(([key, value]) => `${camelToSnake(key)}: ${cssValueToString(value)};`)
    .join('')}
`;

/**
 * Convert CSSObject tokens to CSS FlattenSimpleInterpolation
 * @param tokens Tokens of CSSObjects
 */
export const cssObjectsToCss = <T, K extends keyof T>(tokens: T) =>
  Object.entries(tokens).reduce(
    (retObj, [key, value]) => ({
      ...retObj,
      [key]: cssObjectToCss(value),
    }),
    {} as { [key in K]: FlattenSimpleInterpolation },
  );
