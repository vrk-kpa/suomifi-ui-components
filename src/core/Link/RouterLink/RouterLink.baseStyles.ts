import { css } from 'styled-components';
import { SuomifiTheme } from '../../theme';
import { baseStyles } from '../BaseLink/BaseLink.baseStyles';
import { MarginProps, buildSpacingCSS } from '../../theme/utils/spacing';

export const RouterLinkStyles = (
  theme: SuomifiTheme,
  globalMargins?: MarginProps,
  propMargins?: MarginProps,
) => css`
  ${buildSpacingCSS(globalMargins)}
  ${buildSpacingCSS(propMargins, true)}
  ${baseStyles(theme)}
`;
