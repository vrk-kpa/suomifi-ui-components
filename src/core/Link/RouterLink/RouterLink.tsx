// import React, { Component, ReactNode } from 'react';
import React, { forwardRef, ReactNode } from 'react';
import { default as styled } from 'styled-components';
import { RouterLinkStyles } from './RouterLink.baseStyles';
import { IconChevronRight } from 'suomifi-icons';
import {
  SuomifiThemeProp,
  SuomifiThemeConsumer,
  SpacingConsumer,
} from '../../theme';
import {
  separateMarginProps,
  MarginProps,
  GlobalMarginProps,
} from '../../theme/utils/spacing';
import {
  baseClassName,
  linkClassNames,
  UnderlineVariant,
} from '../BaseLink/BaseLink';
import classnames from 'classnames';
import { HtmlA } from '../../../reset';
import { filterDuplicateKeys } from '../../../utils/common/common';

//
// Source: https://www.benmvp.com/blog/polymorphic-react-components-typescript/
//
// Source: https://github.com/emotion-js/emotion/blob/master/packages/styled-base/types/helper.d.ts
// A more precise version of just React.ComponentPropsWithoutRef on its own
export type PropsOf<
  C extends keyof JSX.IntrinsicElements | React.JSXElementConstructor<any>,
> = JSX.LibraryManagedAttributes<C, React.ComponentPropsWithoutRef<C>>;

type AsProp<C extends React.ElementType> = {
  /**
   * An override of the default HTML tag.
   * Can also be another React component.
   * @default <a>
   */
  asComponent?: C;
};

/**
 * Allows for extending a set of props (`ExtendedProps`) by an overriding set of props
 * (`OverrideProps`), ensuring that any duplicates are overridden by the overriding
 * set of props.
 */
export type ExtendableProps<
  ExtendedProps = {},
  OverrideProps = {},
> = OverrideProps & Omit<ExtendedProps, keyof OverrideProps>;

/**
 * Allows for inheriting the props from the specified element type so that
 * props like children, className & style work, as well as element-specific
 * attributes like aria roles. The component (`C`) must be passed in.
 */
export type InheritableElementProps<
  C extends React.ElementType,
  Props = {},
> = ExtendableProps<PropsOf<C>, Props>;

/**
 * Allows the component's ref type to comply with "asComponent" prop
 * E.g. asComponent = 'button', ref type must be React.Ref<HTMLButtonElement>
 */
type PolymorphicRef<C extends React.ElementType> =
  React.ComponentPropsWithRef<C>['ref'];

/**
 * This type definition exists solely to provide the component with a `forwardedRef` prop which
 * is visible in the Styleguidist API documentation ("ref" is not visible in Styleguidist)
 */
type ForwardedRef<C extends React.ElementType> = {
  /** Alternative for React `ref` attribute. Ref type is derived from the `asComponent` prop value.
   * Example: asComponent = 'button', ref type must be HTMLButtonElement
   */
  forwardedRef?: PolymorphicRef<C>;
};

/**
 * A more sophisticated version of `InheritableElementProps` where
 * the passed in `as` prop will determine which props can be included
 */
export type PolymorphicComponentProps<
  C extends React.ElementType,
  Props = {},
> = InheritableElementProps<C, Props & AsProp<C> & ForwardedRef<C>>;

//
//
//

const routerLinkClassName = `${baseClassName}--router`;

interface Props extends MarginProps {
  /** CSS class for custom styles */
  className?: string;
  /**
   * Text content for the link
   */
  children: ReactNode;
  /**
   * 'initial' | 'hover'
   *
   * Option 'initial' shows underline in link's normal state, and no underline on hover.
   * Option 'hover' shows underline on hover, and no underline in link's normal state.
   *
   * Note: default will be changed from 'hover' to 'initial', so set 'hover' explicitly when necessary.
   *
   * @default hover
   */
  underline?: UnderlineVariant;
}

export type RouterLinkProps<C extends React.ElementType> =
  PolymorphicComponentProps<C, Props>;

const PolymorphicLink = <C extends React.ElementType>(
  props: RouterLinkProps<C> & SuomifiThemeProp,
) => {
  const {
    asComponent,
    globalMargins, // destructured out here to make typescript happy on styled level
    children,
    className,
    smallScreen,
    variant,
    theme,
    underline = 'hover',
    forwardedRef,
    ...rest
  } = props;
  const [_marginProps, passProps] = separateMarginProps(rest);

  const Component = asComponent || HtmlA;

  const classNames = classnames(baseClassName, routerLinkClassName, className, {
    [linkClassNames.linkUnderline]: underline === 'initial',
    [linkClassNames.accent]: variant === 'accent',
    [linkClassNames.small]: smallScreen,
  });

  // If asComponent is included, we assume it can take a normal ref-prop
  if (!!asComponent) {
    return (
      <Component
        className={classNames}
        ref={forwardedRef}
        {...passProps}
        style={{ ...passProps?.style }}
      >
        {variant === 'accent' && (
          <IconChevronRight color={theme.colors.accentBase} />
        )}
        {children}
      </Component>
    );
  }

  // HtmlA (which is rendered by default) exposes a prop called forwardedRef instead
  return (
    <Component
      className={classNames}
      forwardedRef={forwardedRef}
      {...passProps}
      style={{ ...passProps?.style }}
    >
      {children}
    </Component>
  );
};

const StyledRouterLink = styled(
  <C extends React.ElementType>(
    props: RouterLinkProps<C> & SuomifiThemeProp & GlobalMarginProps,
  ) => {
    const { ...passProps } = props;
    return <PolymorphicLink {...passProps} />;
  },
)`
  ${({ theme, globalMargins, ...rest }) => {
    const [marginProps, _passProps] = separateMarginProps(rest);
    const cleanedGlobalMargins = filterDuplicateKeys(
      globalMargins.routerLink,
      marginProps,
    );
    return RouterLinkStyles(theme, cleanedGlobalMargins, marginProps);
  }}
`;

const RouterLinkInner = <C extends React.ElementType = 'a'>(
  props: RouterLinkProps<C>,
  ref?: PolymorphicRef<C>,
) => (
  <SpacingConsumer>
    {({ margins }) => (
      <SuomifiThemeConsumer>
        {({ suomifiTheme }) => (
          <StyledRouterLink
            theme={suomifiTheme}
            globalMargins={margins}
            forwardedRef={ref}
            {...props}
          />
        )}
      </SuomifiThemeConsumer>
    )}
  </SpacingConsumer>
);

// Type assertion is needed to set the function signature with generic type C.
export const RouterLink = forwardRef(RouterLinkInner) as <
  C extends React.ElementType = 'a',
>(
  props: RouterLinkProps<C> & {
    ref?: PolymorphicRef<C>;
  },
) => ReturnType<typeof RouterLinkInner>;

// Because of type assertion the displayName has to be set like this
(RouterLink as React.FC).displayName = 'RouterLink';
