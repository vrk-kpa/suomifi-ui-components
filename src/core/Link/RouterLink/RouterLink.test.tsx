import React from 'react';
import { render } from '@testing-library/react';
import { axeTest } from '../../../utils/test';

import { RouterLink } from './RouterLink';

const TestRouterLink = (
  <RouterLink href="/" data-testid="test-link">
    Hey this is a test
  </RouterLink>
);

const RouterLinkAsButton = (
  <RouterLink asComponent="button">Router link as a button</RouterLink>
);

const RouterLinkVariant = (
  <RouterLink underline="initial">RouterLink variant</RouterLink>
);

const RouterLinkButtonVariant = (
  <RouterLink asComponent="button" underline="initial">
    Router link as button variant
  </RouterLink>
);

test('calling render with the same component on the same container does not remount', () => {
  const { getByTestId, container } = render(TestRouterLink);
  expect(container.firstChild).toMatchSnapshot();
  expect(getByTestId('test-link').textContent).toBe('Hey this is a test');
});

it('should still have the correct styles applied when rendered as something other than <a> ', () => {
  const { container } = render(RouterLinkAsButton);
  expect(container.firstChild).toHaveClass('fi-link--router');
});

describe('margin', () => {
  it('should have margin style from margin prop', () => {
    const { container } = render(
      <RouterLink href="/" margin="xs">
        Link
      </RouterLink>,
    );
    const link = container.querySelector('a');
    expect(link).toHaveStyle('margin: 10px');
  });

  it('should have margin prop overwritten from style prop', () => {
    const { container } = render(
      <RouterLink href="/" margin="xs" style={{ margin: 2 }}>
        Link
      </RouterLink>,
    );
    const link = container.querySelector('a');
    expect(link).toHaveAttribute('style', 'margin: 2px;');
  });
});

test('should not have underline by default', () => {
  const { container } = render(TestRouterLink);
  expect(container.firstChild).not.toHaveClass('fi-link--initial-underline');
});

test('should not have underline be default when rendered as something other than <a>', () => {
  const { container } = render(RouterLinkAsButton);
  expect(container.firstChild).not.toHaveClass('fi-link--initial-underline');
});

test('should have option for underline', () => {
  const { container } = render(RouterLinkVariant);
  expect(container.firstChild).toHaveClass('fi-link--initial-underline');
});

test('should have option for underline when rendered as something other than <a>', () => {
  const { container } = render(RouterLinkButtonVariant);
  expect(container.firstChild).toHaveClass('fi-link--initial-underline');
});

test('should not have basic accessibility issues', axeTest(TestRouterLink));
