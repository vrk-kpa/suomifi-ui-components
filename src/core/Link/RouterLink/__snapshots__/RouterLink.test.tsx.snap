// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`calling render with the same component on the same container does not remount 1`] = `
.c0 {
  line-height: 1.15;
  -ms-text-size-adjust: 100%;
  -webkit-text-size-adjust: 100%;
  background-color: transparent;
  margin: 0;
  padding: 0;
  border: 0;
  box-sizing: border-box;
  font: 100% inherit;
  line-height: 1;
  text-align: left;
  -webkit-text-decoration: none;
  text-decoration: none;
  vertical-align: baseline;
  color: inherit;
  background: none;
  cursor: inherit;
  display: inline;
  -webkit-text-decoration: underline;
  text-decoration: underline;
}

.c0::before,
.c0::after {
  box-sizing: border-box;
}

.c0:hover,
.c0:active,
.c0:focus,
.c0:focus-within {
  color: inherit;
  -webkit-text-decoration: underline;
  text-decoration: underline;
  cursor: pointer;
}

.c1 {
  -webkit-letter-spacing: 0;
  -moz-letter-spacing: 0;
  -ms-letter-spacing: 0;
  letter-spacing: 0;
  -webkit-text-decoration: none;
  text-decoration: none;
  word-break: break-word;
  overflow-wrap: break-word;
  -webkit-font-smoothing: antialiased;
  font-family: 'Source Sans Pro','Helvetica Neue','Arial',sans-serif;
  font-size: 18px;
  line-height: 1.5;
  font-weight: 400;
}

.c1.fi-link {
  color: hsl(212,63%,45%);
  -webkit-text-decoration: none;
  text-decoration: none;
}

.c1.fi-link:hover,
.c1.fi-link:active,
.c1.fi-link:focus,
.c1.fi-link:focus-within {
  color: hsl(212,63%,45%);
}

.c1.fi-link:focus,
.c1.fi-link:focus-within {
  -webkit-text-decoration: none;
  text-decoration: none;
}

.c1.fi-link:focus {
  border-radius: 2px;
  box-shadow: 0 0 0 2px hsl(196,77%,44%);
  outline: 0;
  outline: 3px solid transparent;
}

.c1.fi-link:hover,
.c1.fi-link:active {
  -webkit-text-decoration: underline;
  text-decoration: underline;
}

.c1.fi-link:visited {
  color: hsl(284,36%,45%);
}

.c1.fi-link--initial-underline {
  -webkit-text-decoration: underline;
  text-decoration: underline;
}

.c1.fi-link--initial-underline:focus,
.c1.fi-link--initial-underline:focus-within {
  -webkit-text-decoration: underline;
  text-decoration: underline;
}

.c1.fi-link--initial-underline:hover,
.c1.fi-link--initial-underline:active {
  -webkit-text-decoration: none;
  text-decoration: none;
}

.c1.fi-link.fi-link--accent .fi-link--accent_icon {
  font-size: 16px;
  margin-left: -3px;
  margin-right: 2px;
  -webkit-transform: translateY(0.15em);
  -ms-transform: translateY(0.15em);
  transform: translateY(0.15em);
}

.c1.fi-link.fi-link--small {
  font-size: 16px;
}

<a
  class="c0 fi-link fi-link--router c1"
  data-testid="test-link"
  href="/"
>
  Hey this is a test
</a>
`;
